File: 3319-k-th-largest-perfect-subtree-size-in-binary-tree/3319-k-th-largest-perfect-subtree-size-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right)
 *         : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> ans;
    int dfs(TreeNode* root) {
        if (!root)
            return 0;

        int leftHeight = dfs(root->left);
        int rightHeight = dfs(root->right);

        if (leftHeight == -1 || rightHeight == -1 ||
            leftHeight != rightHeight)
            return -1;
        int currHeight = leftHeight + 1;

        // 2^height - 1
        int size = (1 << currHeight) - 1;
        ans.push_back(size);

        return currHeight;
    }

    int kthLargestPerfectSubtree(TreeNode* root, int k) {
        ans.clear();

        dfs(root);

        sort(ans.begin(), ans.end(), greater<int>());

        if (ans.size() < k)
            return -1;

        return ans[k - 1];
    }
};

File: 3319-k-th-largest-perfect-subtree-size-in-binary-tree/3319-k-th-largest-perfect-subtree-size-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def subtree(self, root):
        if not root:
            return 0
        left = self.subtree(root.left)
        right = self.subtree(root.right)
        if left != right or left == -1 or right == -1:
            return -1
        total = left + right + 1
        self.lst.append(total)
        return total

    def kthLargestPerfectSubtree(self, root: Optional[TreeNode], k: int) -> int:
        self.lst = []
        self.subtree(root)
        self.lst.sort(reverse = True)
        print(self.lst)
        if k > len(self.lst):
            return -1
        return self.lst[k-1]
